test/3des-ecb-test.pl

```diff
@@ -4,4 +4,5 @@
 use warnings;
 use test;
 
-test::cipher("des-ede3", 24, 0);
+# No iv needed, but openssl app version 3 somehow requires one...
```

Comment:
Yea you better check.... Not that they have fixed this in the meantime and now they don't allow to specify an IV for this cipher and fail because of that ;-) 

I do see this in OpenSSL master in apps/enc.c, but this of course depends on the EVP_CIPHER_get_iv_length() to return the correct value.

```
        if (hiv != NULL) {
            int siz = EVP_CIPHER_get_iv_length(cipher);
            if (siz == 0) {
                BIO_printf(bio_err, "warning: iv not used by this cipher\n");
            } else if (!set_hex(hiv, iv, siz)) {
                BIO_printf(bio_err, "invalid hex iv value\n");
                goto end;
            }
        }
        if ((hiv == NULL) && (str == NULL)
            && EVP_CIPHER_get_iv_length(cipher) != 0) {
            /*
             * No IV was explicitly set and no IV was generated.
             * Hence the IV is undefined, making correct decryption impossible.
             */
            BIO_printf(bio_err, "iv undefined\n");
            goto end;
        }

```

Reply:
Well, in the implementation of the default provider you find this:
```
/*
 * NOTE: ECB mode does not use an IV - but existing test code is setting
 * an IV. Fixing this could potentially make applications break.
 */
```
So by trying not to break code they broke our test...

Reply:
I would still suggest report it to OpenSSL. It is clearly a change in behavior compared to OpenSSL 1.1.1, and thus a regression.

Reply:
In OpenSSL 1.1.1 the DES_EDE3 cipher is constructed in evp/e_des3.c:

```
    BLOCK_CIPHER_defs(des_ede3, DES_EDE_KEY, NID_des_ede3, 8, 24, 8, 64,
                  EVP_CIPH_RAND_KEY | EVP_CIPH_FLAG_DEFAULT_ASN1,
                  des_ede3_init_key, NULL, NULL, NULL, des3_ctrl)
```

where the 64 is the IV length.
However, the ECB mode define is the following:

```
#define BLOCK_CIPHER_defs(cname, kstruct, \
                          nid, block_size, key_len, iv_len, cbits, flags, \
                          init_key, cleanup, set_asn1, get_asn1, ctrl) \
....
BLOCK_CIPHER_def_ecb(cname, kstruct, nid, block_size, key_len, flags, \
                     init_key, cleanup, set_asn1, get_asn1, ctrl)
```
It does not pass the iv_len to BLOCK_CIPHER_def_ecb, and BLOCK_CIPHER_def_ecb sets 0 as IV length:

```
#define BLOCK_CIPHER_def_ecb(cname, kstruct, nid, block_size, key_len, \
                             flags, init_key, cleanup, set_asn1, \
                             get_asn1, ctrl) \
BLOCK_CIPHER_def1(cname, ecb, ecb, ECB, kstruct, nid, block_size, key_len, \
                  0, flags, init_key, cleanup, set_asn1, get_asn1, ctrl)
```
This sets the 0 into EVP_CIPHER as iv_len.

Consequently EVP_CIPHER_iv_length() will return 0 for an ECB mode cipher:
```
int EVP_CIPHER_iv_length(const EVP_CIPHER *cipher)
{
    return cipher->iv_len;
}
```

In OpenSSL 3.0 EVP_CIPHER_iv_length is an alias of EVP_CIPHER_get_iv_length.

So clearly, in OpenSSL 3.0 EVP_CIPHER_iv_length  behaves differently than in 1.1.1.